crypto/threads_win.c

```diff
@@ -27,30 +28,362 @@
+
+# define READER_MASK     (((LONG64)1 << READER_SIZE)-1)
+# define ID_MASK         (((LONG64)1 << ID_SIZE)-1)
+# define READER_COUNT(x) (((LONG64)(x) >> READER_SHIFT) & READER_MASK)
+# define ID_VAL(x)       (((LONG64)(x) >> ID_SHIFT) & ID_MASK)
+# define VAL_READER      ((LONG64)1 << READER_SHIFT)
+# define VAL_ID(x)       ((LONG64)x << ID_SHIFT)
+
+struct rcu_qp {
+    volatile LONG64 users;
+};
+
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ * count is the threads current recursion count
+ */
+struct rcu_thr_data {
+    struct rcu_qp* qp;
+    int count;
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_internal {
+    CRYPTO_RCU_LOCK lock; /* Must be first */
+    struct rcu_cb_item *cb_items;
+    uint32_t id_ctr;
+    struct rcu_qp *qp_group;
+    size_t group_count;
+    uint32_t next_to_retire;
+    volatile long int reader_idx;
+    uint32_t current_alloc_idx;
+    uint32_t writers_alloced;
+    CRITICAL_SECTION write_lock;
+    CRITICAL_SECTION alloc_lock;
+    CONDITION_VARIABLE alloc_signal;
+    CRITICAL_SECTION prior_lock;
+    CONDITION_VARIABLE prior_signal;
+};
+
+
+/*
+ * Called on thread exit to free the pthread key
+ * associated with this thread, if any
+ */
+static void free_rcu_thr_data(void *ptr)
+{
+    struct rcu_thr_data *data = ptr;
+
+    assert(data->qp != NULL);
+    OPENSSL_free(data);
+}
+
+
+void CRYPTO_THREAD_rcu_init(void)
+{
+   CRYPTO_THREAD_init_local(&rcu_thr_key, free_rcu_thr_data);
+}
+
+static struct rcu_qp *allocate_new_qp_group(struct rcu_lock_internal *ilock,
+                                            int count)
+{
+    struct rcu_qp *new =
+        OPENSSL_zalloc(sizeof(struct rcu_qp) * count);
+
+    ilock->group_count = count;
+    return new;
+}
+
+CRYPTO_RCU_LOCK *CRYPTO_THREAD_rcu_lock_new(void)
+{
+    struct rcu_lock_internal *new =
+        OPENSSL_zalloc(sizeof(struct rcu_lock_internal));
+
+    if (new == NULL)
+        return NULL;
+
+    InitializeCriticalSection(&new->write_lock);
+    InitializeConditionVariable(&new->alloc_signal);
+    InitializeConditionVariable(&new->prior_signal);
+    InitializeCriticalSection(&new->alloc_lock);
+    InitializeCriticalSection(&new->prior_lock);
+    new->qp_group = allocate_new_qp_group(new, 2);
+    if (new->qp_group == NULL) {
+        OPENSSL_free(new);
+        new = NULL;
+    }
+    return (CRYPTO_RCU_LOCK *)new;
+
+}
+
+void CRYPTO_THREAD_rcu_lock_free(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ldata = (struct rcu_lock_internal *)lock;
+    OPENSSL_free(ldata->qp_group);
+    OPENSSL_free(ldata);
+}
+
+static inline struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+{
+    uint32_t qp_idx;
+    struct rcu_lock_internal *ilock =
+        (struct rcu_lock_internal *)lock;
+
+    /*
+     * get the current qp index
+     */
```

Comment:
Our style for a single line comment is like this:

````
/* get the current qp index */
````

This same comment applies to lots of other places in this PR